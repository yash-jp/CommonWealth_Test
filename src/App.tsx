import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { contributionEdit } from './actions/contribution';
import Contributions from './components/Contributions/Contributions';
import ContributionEdit from './components/ContributionEdit/ContributionEdit';
import { getSelected } from './selectors/contributions';

const App = () => {
  const selected = useSelector(getSelected);
  const dispatch = useDispatch();

  return (
    <>
      <Card>
        <CardHeader title="My Contributions" />
        <CardContent>
          <Contributions />
        </CardContent>
        <CardActions>
          <Button disabled={!selected} onClick={() => dispatch(contributionEdit())}>Edit</Button>
          <Button disabled={!selected}>Cancel</Button>
        </CardActions>
      </Card>
      <ContributionEdit />
    </>
  );
}

export default App;
