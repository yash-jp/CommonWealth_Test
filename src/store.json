{
  "dialogs": {},
  "contributions": [
    {
      "uuid": "6ee2e1b4-ee06-4027-870b-607beb1d89e6",
      "status": "PENDING",
      "tfsa": 500,
      "rrsp": 1000
    },
    {
      "uuid": "d157ad0e-306a-4cb9-a8a6-2d9c84a201da",
      "status": "PROCESSED",
      "tfsa": 500,
      "rrsp": 1000
    },
    {
      "uuid": "cfde9244-4690-498a-8195-08e32fb5814a",
      "status": "PROCESSED",
      "tfsa": 500,
      "rrsp": 1000
    },
    {
      "uuid": "b87330bf-f424-4a61-9e9e-12bc17237eda",
      "status": "PROCESSED",
      "tfsa": 500,
      "rrsp": 1000
    },
    {
      "uuid": "d4bcd4bd-24e5-4b9b-a374-85d6ee5d5869",
      "status": "PROCESSED",
      "tfsa": 500,
      "rrsp": 1000
    },
    {
      "uuid": "6ee2e1b4-ee06-4027-870b-607beb1d89jk",
      "status": "PENDING",
      "tfsa": 500,
      "rrsp": 1000
    }
  ]
}
