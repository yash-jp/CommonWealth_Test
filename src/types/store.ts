import { WithContributions, WithSelectedContribution } from './contribution';
import { WithDialogs } from './dialog';

export type State = WithContributions
  & WithSelectedContribution
  & WithDialogs
  ;
