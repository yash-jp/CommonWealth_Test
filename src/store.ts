import addDays from 'date-fns/addDays';
import subMonths from 'date-fns/subMonths';
import { createStore as reduxCreateStore, applyMiddleware, compose } from 'redux';
// removing this as I want to use thunk middleware, so to use both I need to have composeEnhancers
// import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';

import rootReducer from './reducers';
import preloadedState from './store.json';
import { Contributions } from './types/contribution';
import { State } from './types/store';

// redux-thunk to handle asynchronous actions
import thunk from 'redux-thunk';

// composeEnhancers to use redux-dev-tools
const composeEnhancers = (window && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const transformState = ({ contributions, dialogs }: any, startDate: Date): State => ({
  dialogs,
  selectedContribution: null,
  contributions: contributions.reduce((result: Contributions, { uuid, status, tfsa, rrsp }: any, index: number) => ({
    ...result,
    [uuid]: {
      uuid,
      status,
      tfsa,
      rrsp,
      total: tfsa + rrsp,
      date: subMonths(startDate, index),
    }
  }), {})
});

// instead of co
export const createStore = (startDate: Date) => 
  reduxCreateStore(rootReducer, transformState(preloadedState, startDate), composeEnhancers(
    applyMiddleware(thunk)
  ));

const store = createStore(addDays(new Date(), 15));

export default store;
