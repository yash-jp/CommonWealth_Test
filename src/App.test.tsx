import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import App from './App';
import { Provider } from 'react-redux';

import { createStore } from './store';

const startDate = new Date(2020, 6, 3);

it('renders with no issues', () => {
  const { container } = render(
    <Provider store={createStore(startDate)}>
      <App />
    </Provider>
  );

  expect(container).toMatchSnapshot();
});

it('opens a dialog when pending contribution is edited', () => {
  const { getByText } = render(
    <Provider store={createStore(startDate)}>
      <App />
    </Provider>
  );

  fireEvent.click(getByText('July 3rd, 2020'));
  fireEvent.click(getByText('Edit'));

  expect(document.body).toMatchSnapshot();
});
