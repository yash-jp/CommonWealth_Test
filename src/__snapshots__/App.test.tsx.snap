// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`opens a dialog when pending contribution is edited 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
          >
            My Contributions
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button Mui-selected"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                July 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                $
                1500
                <span
                  class="MuiBadge-root"
                >
                  <span
                    class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
                  >
                    PENDING
                  </span>
                </span>
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                June 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                May 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                April 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                March 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </ul>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Edit Contribution
          </h2>
        </div>
        <form
          action="#"
        >
          <div
            class="MuiDialogContent-root content"
          >
            <div
              class="MuiFormControl-root MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
                data-shrink="true"
              >
                RRSP
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    $
                  </p>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
                  name="rrsp"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
                data-shrink="true"
              >
                TFSA
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    $
                  </p>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
                  name="tfsa"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Cancel
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Accept
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`renders with no issues 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          My Contributions
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              July 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              $
              1500
              <span
                class="MuiBadge-root"
              >
                <span
                  class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
                >
                  PENDING
                </span>
              </span>
            </p>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              June 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              May 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              April 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              March 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Edit
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Cancel
        </span>
      </button>
    </div>
  </div>
</div>
`;
